File: src/paths.ts
/**
 * Resolves OpenClaw file-system paths without depending on the openclaw core package.
 * Reads conventions directly from the file system (~/.openclaw/).
 */
import { execSync } from "node:child_process";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";

export const WORKSPACE_FILES = [
  "AGENTS.md",
  "SOUL.md",
  "TOOLS.md",
  "IDENTITY.md",
  "USER.md",
  "HEARTBEAT.md",
  "BOOTSTRAP.md",
  "MEMORY.md",
] as const;

export type WorkspaceFileName = (typeof WORKSPACE_FILES)[number];

/**
 * All known skill source directories, in precedence order (lowest → highest).
 * Mirrors the merge order in src/agents/skills/workspace.ts:
 *
 *   bundled < managed < personal-agents < project-agents < workspace
 *
 * "plugin" skill dirs are dynamic (resolved from installed plugin manifests at
 * runtime) and cannot be enumerated without the core runtime; excluded here.
 */
export type SkillSource =
  | "bundled"           // <openclaw-pkg>/skills/             shipped with openclaw
  | "managed"           // ~/.openclaw/skills/                installed via `openclaw skills install`
  | "personal-agents"   // ~/.agents/skills/                  personal cross-workspace skills
  | "project-agents"    // <workspace>/.agents/skills/        project-local skills
  | "workspace";        // <workspace>/skills/                workspace-local skills

export interface SkillSourceDir {
  source: SkillSource;
  dir: string;
}

export interface OpenClawPaths {
  /** ~/.openclaw  (or $OPENCLAW_STATE_DIR) */
  stateDir: string;
  /** ~/.openclaw/workspace  (or workspace-<profile>) */
  workspaceDir: string;
  /** ~/.openclaw/openclaw.json */
  configFile: string;
  /**
   * All skill source directories, in ascending precedence order.
   * Higher-precedence sources override lower-precedence ones (same slug wins).
   * Only includes dirs that actually exist on disk.
   */
  skillSources: SkillSourceDir[];
}

/**
 * Try to locate the openclaw package's bundled skills/ directory.
 *
 * Strategy (in order):
 *   1. $OPENCLAW_BUNDLED_SKILLS_DIR env override
 *   2. Walk up from the `openclaw` binary's realpath to find <pkg-root>/skills/
 *   3. `npm root -g` → <npm-global>/openclaw/skills/
 */
function resolveBundledSkillsDir(): string | undefined {
  // 1. env override (same as core)
  const override = process.env.OPENCLAW_BUNDLED_SKILLS_DIR?.trim();
  if (override && fs.existsSync(override)) return override;

  // 2. walk up from openclaw binary realpath
  try {
    const bin = execSync("which openclaw 2>/dev/null", { stdio: ["pipe", "pipe", "ignore"] })
      .toString()
      .trim();
    if (bin) {
      const real = fs.realpathSync(bin);
      let cur = path.dirname(real);
      for (let i = 0; i < 5; i++) {
        const candidate = path.join(cur, "skills");
        if (fs.existsSync(candidate) && fs.existsSync(path.join(cur, "package.json"))) {
          return candidate;
        }
        const next = path.dirname(cur);
        if (next === cur) break;
        cur = next;
      }
    }
  } catch {
    // ignore
  }

  // 3. npm root -g
  try {
    const npmGlobal = execSync("npm root -g 2>/dev/null", {
      stdio: ["pipe", "pipe", "ignore"],
    })
      .toString()
      .trim();
    if (npmGlobal) {
      const candidate = path.join(npmGlobal, "openclaw", "skills");
      if (fs.existsSync(candidate)) return candidate;
    }
  } catch {
    // ignore
  }

  return undefined;
}

/**
 * Resolve all relevant OpenClaw paths from the environment.
 * Mirrors the logic in src/agents/workspace.ts and src/utils.ts.
 */
export function resolvePaths(env: NodeJS.ProcessEnv = process.env): OpenClawPaths {
  const home = env.HOME ?? env.USERPROFILE ?? os.homedir();

  // $OPENCLAW_STATE_DIR overrides the default ~/.openclaw directory
  const stateDir = env.OPENCLAW_STATE_DIR
    ? path.resolve(env.OPENCLAW_STATE_DIR)
    : path.join(home, ".openclaw");

  // Profile-aware workspace directory (mirrors resolveDefaultAgentWorkspaceDir)
  const profile = env.OPENCLAW_PROFILE?.trim();
  const workspaceDirName =
    profile && profile.toLowerCase() !== "default"
      ? `workspace-${profile}`
      : "workspace";
  const workspaceDir = path.join(stateDir, workspaceDirName);

  const allSources: Array<{ source: SkillSource; dir: string }> = [
    { source: "managed",         dir: path.join(stateDir, "skills") },
    { source: "personal-agents", dir: path.join(home, ".agents", "skills") },
    { source: "project-agents",  dir: path.join(workspaceDir, ".agents", "skills") },
    { source: "workspace",       dir: path.join(workspaceDir, "skills") },
  ];

  // bundled: only add if we can find it, inserted at lowest precedence
  const bundledDir = resolveBundledSkillsDir();
  if (bundledDir) {
    allSources.unshift({ source: "bundled", dir: bundledDir });
  }

  // only keep dirs that exist
  const skillSources = allSources.filter(({ dir }) => {
    try { return fs.statSync(dir).isDirectory(); } catch { return false; }
  });

  return {
    stateDir,
    workspaceDir,
    configFile: path.join(stateDir, "openclaw.json"),
    skillSources,
  };
}

File: src/max_context_share/paths.py
r"""OpenClaw file-system path resolution.

Resolves all relevant OpenClaw directories without depending
on the openclaw core package.  Reads conventions directly
from the file system (``~/.openclaw/``).

Attributes:
    WORKSPACE_FILES (tuple[str, ...]): The 8 canonical
        workspace context file names.
"""

from __future__ import annotations

import os
import shutil
import subprocess
from dataclasses import dataclass, field
from pathlib import Path
from typing import List, Literal, Optional

# ── Constants ────────────────────────────────────────────

WORKSPACE_FILES: tuple[str, ...] = (
    "AGENTS.md",
    "SOUL.md",
    "TOOLS.md",
    "IDENTITY.md",
    "USER.md",
    "HEARTBEAT.md",
    "BOOTSTRAP.md",
    "MEMORY.md",
)

# Skill source types in ascending precedence.
SkillSource = Literal[
    "bundled",
    "managed",
    "personal-agents",
    "project-agents",
    "workspace",
]


# ── Data classes ─────────────────────────────────────────

@dataclass(frozen=True)
class SkillSourceDir:
    r"""A pairing of a skill source type with its directory.

    Attributes:
        source (SkillSource): The source category.
        dir (Path): Absolute path to the directory.
    """

    source: SkillSource
    dir: Path


@dataclass(frozen=True)
class OpenClawPaths:
    r"""Resolved OpenClaw paths for the current environment.

    Attributes:
        state_dir (Path): ``~/.openclaw``
            (or ``$OPENCLAW_STATE_DIR``).
        workspace_dir (Path): ``~/.openclaw/workspace``
            (or ``workspace-<profile>``).
        config_file (Path): ``~/.openclaw/openclaw.json``.
        skill_sources (list[SkillSourceDir]): All skill
            source directories in ascending precedence,
            filtered to those that exist on disk.
    """

    state_dir: Path
    workspace_dir: Path
    config_file: Path
    skill_sources: List[SkillSourceDir] = field(
        default_factory=list,
    )


# ── Public API ───────────────────────────────────────────

def resolve_paths(
    env: Optional[dict[str, str]] = None,
) -> OpenClawPaths:
    r"""Resolve all relevant OpenClaw paths from the env.

    Mirrors the logic in the TypeScript ``resolvePaths()``.

    Args:
        env: Environment variable mapping.  Defaults to
            ``os.environ`` when *None*.

    Returns:
        An ``OpenClawPaths`` instance with all directories
        resolved and filtered to those that exist.
    """
    if env is None:
        env = dict(os.environ)

    home = Path(
        env.get("HOME")
        or env.get("USERPROFILE")
        or Path.home().as_posix()
    )

    # $OPENCLAW_STATE_DIR overrides ~/.openclaw
    state_dir_raw = env.get("OPENCLAW_STATE_DIR", "")
    state_dir = (
        Path(state_dir_raw).resolve()
        if state_dir_raw.strip()
        else home / ".openclaw"
    )

    # Profile-aware workspace directory
    profile = (env.get("OPENCLAW_PROFILE") or "").strip()
    ws_name = (
        f"workspace-{profile}"
        if profile and profile.lower() != "default"
        else "workspace"
    )
    workspace_dir = state_dir / ws_name

    # All candidate skill source dirs (ascending precedence)
    all_sources: list[tuple[SkillSource, Path]] = []

    # bundled: lowest precedence, only if found
    bundled = resolve_bundled_skills_dir(env)
    if bundled is not None:
        all_sources.append(("bundled", bundled))

    all_sources.extend([
        ("managed", state_dir / "skills"),
        ("personal-agents", home / ".agents" / "skills"),
        (
            "project-agents",
            workspace_dir / ".agents" / "skills",
        ),
        ("workspace", workspace_dir / "skills"),
    ])

    # Keep only dirs that actually exist
    skill_sources = [
        SkillSourceDir(source=src, dir=d)
        for src, d in all_sources
        if d.is_dir()
    ]

    return OpenClawPaths(
        state_dir=state_dir,
        workspace_dir=workspace_dir,
        config_file=state_dir / "openclaw.json",
        skill_sources=skill_sources,
    )


def resolve_bundled_skills_dir(
    env: Optional[dict[str, str]] = None,
) -> Optional[Path]:
    r"""Locate the openclaw package's bundled skills dir.

    Strategy (in order):
      1. ``$OPENCLAW_BUNDLED_SKILLS_DIR`` env override
      2. Walk up from the ``openclaw`` binary realpath
      3. ``npm root -g`` fallback

    Args:
        env: Environment mapping. Defaults to ``os.environ``.

    Returns:
        Path to the bundled skills directory, or *None*.
    """
    if env is None:
        env = dict(os.environ)

    # 1. env override
    override = (
        env.get("OPENCLAW_BUNDLED_SKILLS_DIR", "") or ""
    ).strip()
    if override:
        p = Path(override)
        if p.exists():
            return p

    # 2. walk up from openclaw binary realpath
    bin_path = shutil.which("openclaw")
    if bin_path:
        try:
            real = Path(bin_path).resolve()
            cur = real.parent
            for _ in range(5):
                candidate = cur / "skills"
                pkg_json = cur / "package.json"
                if candidate.exists() and pkg_json.exists():
                    return candidate
                parent = cur.parent
                if parent == cur:
                    break
                cur = parent
        except OSError:
            pass

    # 3. npm root -g
    try:
        result = subprocess.run(
            ["npm", "root", "-g"],
            capture_output=True,
            text=True,
            timeout=5,
        )
        npm_global = result.stdout.strip()
        if npm_global:
            candidate = Path(npm_global) / "openclaw" / "skills"
            if candidate.exists():
                return candidate
    except (OSError, subprocess.TimeoutExpired):
        pass

    return None
